core/src/main/java/com/linecorp/armeria/server/Server.java

```diff
@@ -473,6 +479,42 @@ public void reconfigure(ServerConfigurator serverConfigurator) {
         config.serviceConfigs().forEach(cfg -> ServiceCallbackInvoker.invokeServiceAdded(cfg, cfg.service()));
     }
 
+    /**
+     * Registers a JVM shutdown hook that closes this {@link Server} when the current JVM terminates.
+     */
+    public CompletableFuture<Void> closeOnShutdown() {
+        return closeOnShutdown(null);
+    }
+
+    /**
+     * Registers a JVM shutdown hook that closes this {@link Server} when the current JVM terminates.
+     *
+     * @param whenClosing the {@link Runnable} will be run before closing this {@link Server}
+     */
+    public CompletableFuture<Void> closeOnShutdown(@Nullable Runnable whenClosing) {
+        final CompletableFuture<Void> future = new CompletableFuture<>();
+        Runtime.getRuntime().addShutdownHook(THREAD_FACTORY.newThread(() -> {
```

Comment:
Give that both `ClientFactory` and `Server` are `AsyncCloseable`, how about extracting the code that registers the shutdown hook and keeps the tasks in a `LinkedHashMap` into a new class such as `AsyncCloseableShutdownHooks`? e.g.

```
public CompletableFuture<Void> closeOnShutdown(@Nullable Runnable whenClosing) {
    return AsyncCloseableShutdownHooks.add(this, whenClosing);
}
```